### ISSUE TITLE
Implement macOS dispatch semaphores

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/miri#2589]
https://github.com/rust-lang/rust/pull/102773 started using macOS dispatch semaphores for thread parking on macOS. We should implement those APIs in Miri. (Right now a `cfg(miri)` makes miri use a pthread-based fallback implementation.)

[These seem to be the API docs](https://developer.apple.com/documentation/dispatch/dispatch_semaphore?language=objc).

**Comment by thomcc** [comment]
Yes, those are the docs. Feel free to ping me if you have questions.

Note that its lifetime is managed by `dispatch_retain`/`dispatch_release` calls, which manage an atomic reference count. We only use `dispatch_release` in std patch (since the object returned by `dispatch_create` has a retain count of 1 initially), and I don't see a reason in the near term we'd use `dispatch_retain` for anything, but I'd advise against implementing this in a way that adding the refcount later would require major hacks or something, not that that seems likely.

**Comment by RalfJung** [comment]
https://github.com/rust-lang/rust/pull/122408 may switch the macOS primitives to yet another OS API, so I am not sure if it would currently be worth it to implement dispatch semaphores.

**Comment by RalfJung** [comment]
@joboet @thomcc looks like https://github.com/rust-lang/rust/pull/122408 isn't moving anywhere so maybe we should implement dispatch semaphores after all, so we can get rid of a `cfg(miri)`?
